substrate/frame/executive/src/lib.rs

```diff
@@ -375,12 +377,6 @@ where
 		// Nothing should modify the state after the migrations ran:
 		let _guard = StorageNoopGuard::default();
 
-		// The state must be decodable:
-		if checks.any() {
-			let res = AllPalletsWithSystem::try_decode_entire_state();
-			Self::log_decode_result(res)?;
-		}
```

Comment:
Why do we have the `checks` parameter? I mean we could also just add a new variant to check? So no new runtime api function.

Reply:
`UpgradeCheckSelect` is currently an enum:

```rust
pub enum UpgradeCheckSelect {
	None,
	All,
	PreAndPost,
	TryState,
}
```

So if we want to make `try_decode_entire_state` configurable by that arg, I think we'd want to update it to a struct something like this:

```rust
struct UpgradeCheckSelect {
	pre_and_post: bool,
	try_state: Select,
	decode_entire_state: bool	
}
```

Otherwise we'd need to add an enum variant for every combination of checks the developer wants to run. Let me refactor to do that..

I there there still is value to de-coupling the try-state and decode-entire-state checks from on-runtime-upgrade and execute-block methods, but that requires a bigger change and can leave it for a future PR. Let me work on the struct refactor first.

Reply:
Let me answer this question in 2 parts.

1. Suggest how we could add it to checks parameter, but why I don't think this is the best solution
2. Suggest an alternative

## 1. How we could add it to `checks` 

`UpgradeCheckSelect` is currently an enum:

```rust
pub enum UpgradeCheckSelect {
	None,
	All,
	PreAndPost,
	TryState,
}
```

To keep the enum, we'd need to add a new variant for every combination of checks the user may want to run. Which is obviously not scalable or a good solution.

Instead, I think we'd want to change it to a struct something like this:

```rust
struct UpgradeCheckSelect {
	pre_and_post: bool,
	try_state: Select,
	decode_entire_state: bool	
}
```

However, I don't think this is optimal for 2 reasons: 

1. It's a breaking change every time we want to add a new `try-runtime` task
2. It's inflexible in the sense that we can only run checks in combination with an `on_runtime_upgrade` check or an `execute_block` check. What if we just want to decode entire state? or run try-state checks?
3. We need to clutter the configuration to both `try_on_runtime_upgrade` and `try_execute_block` with the check options

## 2. Proposed alternative 

1. Add a new `enum` describing all the atomic `try-runtime` related tasks, something like
```rust
pub enum TryRuntimeTask {
	OnRuntimeUpgrade(pre_and_post: bool),
	ExecuteBlock(block: Block, state_root_check: bool, signature_check: bool),
	TryState(TryStateSelect),
	TryDecodeEntireState
}
```

4. Remove all `try-state` and `try-decode-entire-state` logic from `try_on_runtime_upgrade` and `try_execute_block` entirely, instead create dedicated methods  for running those checks `try_state` and `try_decode_entire_state`.

2. Create a new `Executive` method and runtime API that accepts a vec of `TryRuntimeTasks`, and executes them in order

```rust
// runtime api
fn execute_try_runtime_tasks(tasks: Vec<TryRuntimeTask>) -> (Weight, Weight) {
	Executive::execute_try_runtime_tasks(tasks)
}

// executive method
pub fn execute_try_runtime_tasks(Vec<TryRuntimeTasks>) -> (Weight, Weight) {
	let agg_weight = Weight::zero();
	for task in tasks {
		match task {
			TryRuntimeTask::OnRuntimeUpgrade(pre_and_post) => {
				let weight = Self::try_on_runtime_upgrade(pre_and_post);
				agg_weight.saturating_acc(weight);
			}
			TryRuntimeTask::ExecuteBlock(block, state_root_check, signature_check) => {
				let weight = Self::try_execute_block(block, state_root_check, signature_check, select).unwrap();
				agg_weight.saturating_acc(weight);
			}
			TryRuntimeTask::TryState(try_state_select) {
				Self::try_state(try_state_select);
			}
			TryRuntimeTask::TryDecodeEntireState {
				Self::try_decode_entire_state();
			}
		}
	}
	(weight, BlockWeights::get().max_block)
}
```

This allows `try_on_runtime_upgrade` and `try_execute_block` to not be concerned about *what* checks to run or *when* (before or after) to run them or *how* (configuration) to run them. 

The developer instead, with ultimate flexibility, just specifies a Vec like `vec![OnRuntimeUpgrade(OnRuntimeUpgradeConfig), TryState(TryStateConfig), TryDecodeEntireState]` describing what they want to run in what order.

Once this is implemented, we can add a deprecation notice to the current try-runtime runtime APIs and eventually remove them.

Reply:
Okay ty for the writeup. 

> It's a breaking change every time we want to add a new try-runtime task

Your second proposal has the same issues ;) I don't think in general that this is such a problem, because these are development apis that don't need to stay stable or we need to support them over X years.

> It's inflexible in the sense that we can only run checks in combination with an on_runtime_upgrade check or an execute_block check. What if we just want to decode entire state? or run try-state checks?

Not sure why it also requires the `on_runtime_upgrade` check. However, it will always require that we run the runtimes upgrades. You always need to run the upgrades as the new code you are running could be incompatible to the old runtime that created the state. This actually shows a "flaw" in your current pr here, because you don't run the upgrades before.

Reply:
> Your second proposal has the same issues ;) I don't think in general that this is such a problem, because these are development apis that don't need to stay stable or we need to support them over X years.

If it's an enum then I think we can add new variants without it being a breaking change to the caller (e.g. try-runtime-cli)? Since `try-runtime-cli` will just construct the Vec with enums it knows about. 

You're right though not a huge deal, and we likely won't change it frequently so not a huge deal.. 

Do you feel I should proceed with the struct approach (described in 1.) then?

Reply:
> If it's an enum then I think we can add new variants without it being a breaking change to the caller (e.g. try-runtime-cli)? Since `try-runtime-cli` will just construct the Vec with enums it knows about.

Yeah that is a good point. 

Maybe we could also just introduce a bitflag, assuming we don't need to pass some data. However, if we need to pass some data, I would probably use your enum approach given your reasoning above.